python/cudf/cudf/core/scalar.py

```diff
@@ -0,0 +1,344 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+import numpy as np
+
+from cudf._lib.scalar import DeviceScalar, _is_null_host_scalar
+from cudf.core.column.column import ColumnBase
+from cudf.core.index import Index
+from cudf.core.series import Series
+from cudf.utils.dtypes import (
+    get_allowed_combinations_for_operator,
+    to_cudf_compatible_scalar,
+)
+
+
+class Scalar(object):
+
+    _host_value = None
+    _host_dtype = None
```

Comment:
Why is `dtype` qualified with `host_`?

Reply:
So, there are technically two `dtype` attributes we are concerned with. The first is the one that we get from creating a host scalar, and ends up being the dtype that `_preprocess_host_value` figures out for the incoming host scalar. When `cudf.Scalar.device_value` is called, this dtype is used to pick the correct subclass of `cudf::scalar` to allocate and copy the host value into. 

The second dtype is the `dtype` property of the `DeviceScalar` object. This property actually inspects the libcudf c++ type of the device scalar being pointed to, and then maps that to a numpy type via a dict, without caching anything. I am not 100% sure if this process incurs a copy, in this code I have assumed it *does* - and as such, it's written to do a normal device to host sync if necessary and then read the dtype from the scalar rather than accessing the `DeviceScalar`'s dtype attribute. 

Since they seemed distinct in this way I labeled them differently. Although if I'm wrong and accessing `DeviceScalar.dtype` doesn't incur a copy - then `cudf.Scalar.dtype` could just forward to `DeviceScalar.dtype`.